File: app/src/main/java/com/waracle/androidtest/DataSource/ImageSources.java
package com.waracle.androidtest.DataSource;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.waracle.androidtest.ImageLoader;
import com.waracle.androidtest.MainApplication;
import com.waracle.androidtest.R;
import com.waracle.androidtest.StreamUtils;

import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.CopyOnWriteArrayList;

public class ImageSources implements DataSources<Bitmap> {

    private long startTimeStamp;
    private String url;
    private byte[] byteData;
    private Bitmap bitmap;
    private ConcurrentLinkedQueue<WeakReference<DataListeners<Bitmap>>> imageSourceListener = new ConcurrentLinkedQueue<>();

//    private CopyOnWriteArrayList<WeakReference<DataListeners<Bitmap>>> imageSourceListener = new CopyOnWriteArrayList<>();


    public String getUrl() {
        return url;
    }

    @Override
    public ConcurrentLinkedQueue<WeakReference<DataListeners<Bitmap>>> getDataSourceListeners() {
        return imageSourceListener;
    }

    public ImageSources setUrl(String url) {
        this.url = url;
        return this;
    }

    private long getTimeStamp() {
        return startTimeStamp;
    }

    @Override
    public void run() {
        if (bitmap == null ||
                getTimeStamp() > MainApplication.getDataSource().getTimeToCache() + getTimeStamp()) {
            startTimeStamp = System.currentTimeMillis();
            byte[] bytedata = new byte[0];
            try {
                bytedata = loadImageData(url);
            } catch (IOException e) {
                e.printStackTrace();
            }
            if (bytedata != null) {
                bitmap = convertToBitmap(bytedata);
            }
            if (bitmap == null) {
                bitmap = ImageLoader.getBitmapFromVectorDrawable(MainApplication.getInstance(), R.drawable.ic_block_black_24dp);
            }
        }

        for(WeakReference<DataListeners<Bitmap>> dataListener: imageSourceListener){
            if (dataListener.get() != null) {
                dataListener.get().onDataRetrieved(bitmap);
            }
            imageSourceListener.remove(dataListener);
        }
    }

    private byte[] loadImageData(String url) throws IOException {
        InputStream inputStream = null;
        HttpURLConnection.setFollowRedirects(true);
        HttpURLConnection connection = null;
        try {
            connection = (HttpURLConnection) new URL(url).openConnection();
            if (connection.getResponseCode() == 301) {
                loadImageData(connection.getHeaderField("Location"));
            } else {
                try {
                    // Read data from workstation
                    inputStream = connection.getInputStream();
                } catch (IOException e) {
                    // Read the error from the workstation
                    inputStream = connection.getErrorStream();
                }
                byteData = StreamUtils.readUnknownFully(inputStream);
                // Can you think of a way to make the entire
                // HTTP more efficient using HTTP headers??=
            }
        } finally {
            // Close the input stream if it exists.
            StreamUtils.close(inputStream);

            // Disconnect the connection
            if (connection != null) {
                connection.disconnect();
            }
        }
        return byteData;
    }

    private static Bitmap convertToBitmap(byte[] data) {
        return BitmapFactory.decodeByteArray(data, 0, data.length);
    }
}

File: app/src/main/java/com/waracle/androidtest/dataSource/ImageSources.kt
package com.waracle.androidtest.dataSource

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.waracle.androidtest.ImageLoader
import com.waracle.androidtest.MainApplication
import com.waracle.androidtest.R
import java.io.IOException
import java.lang.ref.WeakReference
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ConcurrentLinkedQueue

class ImageSources : DataSources{

    private var byteData: ByteArray? = null
    private var startTimeStamp:Long = 0
    private var url:String = ""
    private var bitmap: Bitmap? = null
    private var imageSourceListener:ConcurrentLinkedQueue<WeakReference<DataSources.DataListeners<Any>>> = ConcurrentLinkedQueue()



    override fun getUrl(): String {
        return url
    }

    fun setUrl(url: String): ImageSources {
        this.url = url
        return this
    }

    private fun getTimeStamp(): Long {
        return startTimeStamp
    }

    override fun getDataSourceListeners(): ConcurrentLinkedQueue<WeakReference<DataSources.DataListeners<Any>>> {
        return imageSourceListener
    }

    override suspend fun run() {
        if(bitmap == null ||
                getTimeStamp() > MainApplication.getDataSource().getTimeToCache() + getTimeStamp()){
            startTimeStamp = System.currentTimeMillis()
            val byteData = loadImageData(url)
            if (byteData != null) {
                bitmap = convertToBitmap(byteData)
            }
            if (bitmap == null) {
                bitmap = ImageLoader.getBitmapFromVectorDrawable(MainApplication.getInstance(), R.drawable.ic_block_black_24dp)
            }
        }

        while (imageSourceListener.peek() != null){
            val dataListener = imageSourceListener.poll()
            if (dataListener?.get() != null){
                dataListener.get()?.onDataRetrieved(bitmap!!)
            }
        }
    }

    @Throws(IOException::class)
    private fun loadImageData(url:String):ByteArray?{
        HttpURLConnection.setFollowRedirects(true)
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connect()
        if (connection.responseCode == 301) {
            loadImageData(connection.getHeaderField("Location"))
        }else{
            try {
                byteData = connection.inputStream.readBytes()
                connection.inputStream.close()
            }catch (e:IOException){
            }finally{
                connection.disconnect()
            }
        }
        return byteData
    }

    private fun convertToBitmap(data: ByteArray): Bitmap? {
        return BitmapFactory.decodeByteArray(data, 0, data.size)
    }
}